### PR TITLE
Remove GUI and Update Docs

### LINKED ISSUE #45: Proposal: Get Rid of Gradio Interface
I spent some time playing with the Gradio interface for the axolotl trainer.  Here are my impressions:

- It doesn't add much in the way of ease of use.  The dataset is pre-baked, and you still need to carefully fill out a `config.yaml` file.  
- We recommend the GUI in the README as the beginner option and the CLI as the `advanced option`.  I would argue that the GUI is more confusing as you need to create a deployment backend as well as the Gradio front end.  The only thing the GUI offers you is to press a button to train, but this is not good enough.  
- The data viewers add no value IMO as they are not formatting the data (and the form factor isn't going to work for many real-world datasets).
- The gradio app has bugs in the inference tab, the model fails to load and config/data are not shown.  

My suggestion is to get rid of the gradio app as it expands the surface area of this tutorial and application in ways that don't seem worth it.  That way, we can have a more focused experience but also make this code more maintainable.  I think deleting that part of the code will let modal really shine.

cc: @charlesfrye

### PR DESCRIPTION
**Pull request by hamelsmu**
This PR is in lieu of #43.  I also am removing the GUI component per discussion [in slack](https://modallabscommunity.slack.com/archives/C06SE4HQW93/p1713054567786809?thread_ts=1713052984.720129&cid=C06SE4HQW93)

This PR closes #45 

(reopening this from a branch instead of a fork).

### CODE REVIEW

**Comment by hamelsmu** [comment]
Ok looks like an error with `wandb`  ... 👀

**Comment by mwaskom** [comment]
Can we please not vary parameters across base models where we don't mean to communicate that the parameterizations are model specific?. i.e. if we're going to switch to deepspeed zero1 let's do it for all of the models not just mistral.

**Review by mwaskom — COMMENTED** [review]
Some nits on the new readme text.

**Comment by hamelsmu** [comment]
@mwaskom about your question

> Can we please not vary parameters across base models where we don't mean to communicate that the parameterizations are model specific?. i.e. if we're going to switch to deepspeed zero1 let's do it for all of the models not just mistral.

I checked and all the other example configs in this repo either do not use any deepspeed config, or are mixtral 8x8b parameter models which are too big to use DS Zero 1.   DeepSpeed parameterizations are indeed model specific IMO in that they are specific to the model's size relative to the GPUs being used.  

- [codellama](https://github.com/modal-labs/llm-finetuning/blob/main/config/codellama.yml):  No DeepSpeed config was present
- [llama-2](https://github.com/modal-labs/llm-finetuning/blob/main/config/llama-2.yml): No DS config was present
- [mistral](https://github.com/modal-labs/llm-finetuning/blob/main/config/mistral.yml) already had a DS config, but I think it was incorrect IMO that is why I changed it.  I don't think we need DS Z1 and all the additional communication overhead for a 7B model on a H100 GPU.  
- mixtral  (there are two configs) - they are both ok in that they are using DS Z2
- [pythia](https://github.com/modal-labs/llm-finetuning/blob/main/config/pythia.yml) - no DS config was present 

I was inferring from this prior structure that we didn't always assume multi-gpu training and its just the example in the README.  

Let me know if I'm misunderstanding or missing something!  cc: @charlesfrye

File: README.md [hunk 1]
```diff
@@ -33,47 +36,74 @@ Follow the steps to quickly train and test your fine-tuned model:
...
-modal run --detach src.train --config=config/mistral.yml --data=data/sqlqa.jsonl
-```
+    ```bash
+    modal run --detach src.train --config=config/mistral.yml --data=data/sqlqa.jsonl
+    ```
+
+Some important caveats about the `train` command:
+
+- The `--data` flag is used to pass your dataset to axolotl. This dataset is then written to the `datasets.path` as specified in your config file. If you alraedy have a dataset at `datasets.path`, you must be careful to also pass the same path to `--data` to ensure the dataset is correctly loaded.
+- Unlike axolotl, you cannot pass additional flags to the `train` command. However, you can specify all your desired options in the config file instead.
```
[the remark this answers is not in the document]
Comment (hamelsmu) [reply README.md:52]: My observations:

- I think this repo is pretty difficult to reason about if you aren't familiar with axolotl IMO.  Like what are these configs?  How does it work?  How are my prompts assembled exactly?  What does the dataset format need to be?  Are there other dataset formats?  How do I check the prompt construction? etc. I was actually assuming that the user is indeed familiar with axolotl.  
- If you are very familiar with `axoltol`, this `--data` flag was really confusing to me, because a key parameter in my config that I am used to using is being completely ignored with an extra layer of indirection.  I actually got stuck on this personally as an experienced axolotl user, so I found the need to provide these two caveats. 

cc: @charlesfrye @winglian curious what you think

File: README.md [hunk 2]
```diff
@@ -33,47 +36,74 @@ Follow the steps to quickly train and test your fine-tuned model:
...
+Our quickstart example trains a 7B model on a text-to-SQL dataset as a proof of concept (it takes just a few minutes). It uses DeepSpeed ZeRO stage 1 to use data parallelism across 2 H100s. Inference on the fine-tuned model displays conformity to the output structure (`[SQL] ... [/SQL]`). To achieve better results, you would need to use more data! Refer to the full development section below.
+
+> [!TIP]
+> DeepSpeed ZeRO-1 is not the best choice if your model doesn't comfortably fit on a single GPU. For larger models, we recommend DeepSpeed Zero stage 3 instead by changing the `deepspeed` configuration path.  Modal mounts the [`deepspeed_configs` folder](https://github.com/OpenAccess-AI-Collective/axolotl/tree/main/deepspeed_configs) from the `axolotl` repository.  You reference these configurations in your `config.yml` like so: `deepspeed: /root/axolotl/deepspeed_configs/zero3_bf16.json`.  If you need to change these standard configurations, you will need to modify the `train.py` script to load your own custom deepspeed configuration.
```
[the remark this answers is not in the document]
Comment (hamelsmu) [reply README.md:65]: I don't like having a mix of config vs CLI where the config is ignored.  I would say have the user supply a DS config and just reference it in the config as normal.

The key benefit of axolotl is that you can share your config files with other users so you can reproduce training runs, and I think such a design would run counter to that purpose

File: config/mistral.yml [hunk 3]
```diff
@@ -46,7 +46,7 @@ wandb_watch:
 wandb_run_id:
 
 gradient_accumulation_steps: 1
-micro_batch_size: 32
+micro_batch_size: 64
```
[the remark this answers is not in the document]
Comment (hamelsmu) [reply config/mistral.yml:49]: I would say perhaps we modify this to be appropriate for the GPU vRAM / Model Size but happy to revert this as I don't think its a big deal in this case

**Comment by mwaskom** [comment]
Hm are we looking in different places?

```
$ git grep zero config
config/codellama.yml:deepspeed: /root/axolotl/deepspeed_configs/zero3_bf16.json
config/mistral.yml:deepspeed: /root/axolotl/deepspeed_configs/zero3_bf16.json
config/mixtral.yml:deepspeed: /root/axolotl/deepspeed_configs/zero2.json
config/mixtral_out_of_box.yml:deepspeed: /root/axolotl/deepspeed_configs/zero2.json
```

That said, already kind of a mess I guess.

**Comment by mwaskom** [comment]
But as I mentioned above, I think that "what deepspeed to use" is a tricky question, i.e. i think that the original intention here was to demonstrate that zero3 *works on modal* even if it's not strictly necessary for the model size that we include in the quickstart. But I appreciate that many users will just pick up that config and run with it. So it's hard — a broader question about the point of this repo, IMO.

**Comment by hamelsmu** [comment]
> Hm are we looking in different places?

I was looking at some old commit ong GitHub sorry about that, not sure how that happened

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwaskom) [reply README.md:65]: My point is that we would need a way to supply a separate deepspeed config to the modal app — since currently we are pointing at a file that lives inside the axolotl docker image. I still don't follow how you could modify the source code here itself to accomplish this.

**Comment by hamelsmu** [comment]
> But I appreciate that many users will just pick up that config and run with it. So it's hard — a broader question about the point of this repo, IMO.

That is an interesting point, perhaps we should huddle and discuss who the think the audience is, what we assume they come with etc.  I probably have a much different mental model.  I think it would be good to understand that so I can make better choices!

**Comment by hamelsmu** [comment]
re: Deepspeed configs

To resolve the gridlock I can just make everything z3 for now; its not like its going to kill anyone even if it is suboptimal

**Comment by mwaskom** [comment]
> its not like its going to kill anyone even if it is suboptimal

Famous last words in LLM finetuning ;)

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (hamelsmu) [reply README.md:65]: Oh I see what you mean, Yeah I meant you need to rebuild the container and mount the new file into it.  I think thats probably way too advanced.  I'll just delete the last sentence

File: README.md [hunk 4]
```diff
@@ -1,22 +1,25 @@
...
-<br>
-  
-- *Deepspeed ZeRO-3* to efficiently shard the base model and training state across multiple GPUs ([more info](https://www.deepspeed.ai/2021/03/07/zero3-offload.html))
-- *Parameter-efficient fine-tuning* via LoRA adapters for faster convergence
-- *Flash attention* for fast and memory-efficient attention during training (note: only works with certain hardware, like A100s)
-- *Gradient checkpointing* to reduce VRAM footprint, fit larger batches and get higher training throughput
-</details>
+ This repository gives the popular [`axolotl`](https://github.com/OpenAccess-AI-Collective/axolotl) fine-tuning library a serverless twist. It uses Modal's serverless infrastructure to run your fine-tuning jobs in the cloud, so you can train your models without worrying about building images or idling expensive GPU VMs. 
 
+ Any application written with Modal, including this one, can be trivially scaled across many GPUs in a reproducible and efficient manner.  This makes any fine-tuning job you prototype with this repository production-ready.
```
[the remark this answers is not in the document]
Comment (hamelsmu) [reply README.md:9]: Addressed this in https://github.com/modal-labs/llm-finetuning/pull/49

File: README.md [hunk 5]
```diff
@@ -1,22 +1,25 @@
 # Fine-tune any LLM in minutes (ft. Mixtral, LLaMA, Mistral)
 
-### Tired of prompt engineering? You've come to the right place.
+This guide will take you from a dataset to using a fine-tuned LLM for inference in a matter of minutes.
```
[the remark this answers is not in the document]
Comment (hamelsmu) [reply README.md:3]: Addressed in https://github.com/modal-labs/llm-finetuning/pull/49

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (hamelsmu) [reply README.md:65]: Adressed in https://github.com/modal-labs/llm-finetuning/pull/49

File: config/mistral.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (hamelsmu) [reply config/mistral.yml:49]: reverted in https://github.com/modal-labs/llm-finetuning/pull/49

File: README.md [hunk 6]
```diff
@@ -33,47 +36,74 @@ Follow the steps to quickly train and test your fine-tuned model:
...
-```
+    ```bash
+    modal run --detach src.train --config=config/mistral.yml --data=data/sqlqa.jsonl
+    ```
+
+Some important caveats about the `train` command:
+
+- The `--data` flag is used to pass your dataset to axolotl. This dataset is then written to the `datasets.path` as specified in your config file. If you alraedy have a dataset at `datasets.path`, you must be careful to also pass the same path to `--data` to ensure the dataset is correctly loaded.
+- Unlike axolotl, you cannot pass additional flags to the `train` command. However, you can specify all your desired options in the config file instead.
+- This example training script is opinionated in order to make it easy to get started.  For example, a LoRA adapter is used and merged into the base model after training. This merging is currently hardcoded into the `train.py` script.  You will need to modify this script if you do not wish to fine-tune using a LoRA adapter.
```
[the remark this answers is not in the document]
Comment (hamelsmu) [reply README.md:53]: Tracking this here https://github.com/modal-labs/llm-finetuning/issues/50
